File: backend/app/services/admin_service.py
from datetime import datetime, timedelta
from typing import Optional, Tuple, Dict, Any
from sqlalchemy import func, and_, or_
from app.database.db import db
from app.models.user import User
from app.models.product import Product
from app.models.order import Order
from app.models.order_item import OrderItem
from app.models.payment import Payment


class AdminService:
    """Admin service for dashboard statistics and management"""

    @staticmethod
    def get_dashboard_stats() -> Dict[str, Any]:
        """
        Get dashboard statistics

        Returns:
            Dictionary with various statistics
        """
        # Get date ranges
        today = datetime.utcnow().date()
        thirty_days_ago = today - timedelta(days=30)
        seven_days_ago = today - timedelta(days=7)

        # Total revenue
        total_revenue = db.session.query(func.sum(Order.total)).filter(
            Order.payment_status == 'paid'
        ).scalar() or 0

        # Revenue this month
        month_revenue = db.session.query(func.sum(Order.total)).filter(
            and_(
                Order.payment_status == 'paid',
                Order.created_at >= thirty_days_ago
            )
        ).scalar() or 0

        # Total orders
        total_orders = Order.query.count()

        # Orders today
        orders_today = Order.query.filter(
            func.date(Order.created_at) == today
        ).count()

        # Pending orders
        pending_orders = Order.query.filter(
            Order.status.in_(['pending', 'processing'])
        ).count()

        # Total products
        total_products = Product.query.filter(Product.is_active == True).count()

        # Low stock products
        low_stock_count = Product.query.filter(
            and_(
                Product.stock_quantity > 0,
                Product.is_active == True
            )
        ).all()
        low_stock_count = len([p for p in low_stock_count if p.is_low_stock])

        # Out of stock products
        out_of_stock = Product.query.filter(
            and_(
                Product.stock_quantity == 0,
                Product.is_active == True
            )
        ).count()

        # Total customers
        total_customers = User.query.filter(User.role == 'customer').count()

        # New customers this week
        new_customers = User.query.filter(
            and_(
                User.role == 'customer',
                User.created_at >= seven_days_ago
            )
        ).count()

        # Recent orders (last 10)
        recent_orders = Order.query.order_by(Order.created_at.desc()).limit(10).all()

        # Top selling products (last 30 days)
        top_products = db.session.query(
            OrderItem.product_id,
            Product.name,
            func.sum(OrderItem.quantity).label('total_sold')
        ).join(
            Order, OrderItem.order_id == Order.id
        ).join(
            Product, OrderItem.product_id == Product.id
        ).filter(
            and_(
                Order.created_at >= thirty_days_ago,
                Order.payment_status == 'paid'
            )
        ).group_by(
            OrderItem.product_id, Product.name
        ).order_by(
            func.sum(OrderItem.quantity).desc()
        ).limit(5).all()

        # Sales chart data (last 7 days)
        sales_chart = []
        for i in range(7):
            date = today - timedelta(days=i)
            daily_sales = db.session.query(func.sum(Order.total)).filter(
                and_(
                    func.date(Order.created_at) == date,
                    Order.payment_status == 'paid'
                )
            ).scalar() or 0

            sales_chart.append({
                'date': date.isoformat(),
                'sales': float(daily_sales)
            })

        sales_chart.reverse()  # Show oldest to newest

        return {
            'revenue': {
                'total': float(total_revenue),
                'this_month': float(month_revenue)
            },
            'orders': {
                'total': total_orders,
                'today': orders_today,
                'pending': pending_orders
            },
            'products': {
                'total': total_products,
                'low_stock': low_stock_count,
                'out_of_stock': out_of_stock
            },
            'customers': {
                'total': total_customers,
                'new_this_week': new_customers
            },
            'recent_orders': [order.to_dict() for order in recent_orders],
            'top_products': [
                {
                    'product_id': p.product_id,
                    'name': p.name,
                    'total_sold': int(p.total_sold)
                }
                for p in top_products
            ],
            'sales_chart': sales_chart
        }

    @staticmethod
    def get_all_orders(
        page: int = 1,
        per_page: int = 20,
        status: str = None,
        payment_status: str = None,
        start_date: str = None,
        end_date: str = None
    ) -> Dict[str, Any]:
        """
        Get all orders with filters and pagination

        Args:
            page: Page number
            per_page: Items per page
            status: Filter by order status
            payment_status: Filter by payment status
            start_date: Filter by start date (YYYY-MM-DD)
            end_date: Filter by end date (YYYY-MM-DD)

        Returns:
            Dictionary with orders and pagination info
        """
        query = Order.query

        # Apply filters
        if status:
            query = query.filter(Order.status == status)

        if payment_status:
            query = query.filter(Order.payment_status == payment_status)

        if start_date:
            try:
                start = datetime.fromisoformat(start_date)
                query = query.filter(Order.created_at >= start)
            except ValueError:
                pass

        if end_date:
            try:
                end = datetime.fromisoformat(end_date)
                # Add one day to include the end date
                end = end + timedelta(days=1)
                query = query.filter(Order.created_at < end)
            except ValueError:
                pass

        # Order by created_at descending
        query = query.order_by(Order.created_at.desc())

        # Paginate
        pagination = query.paginate(
            page=page,
            per_page=per_page,
            error_out=False
        )

        return {
            'orders': [order.to_dict() for order in pagination.items],
            'total': pagination.total,
            'pages': pagination.pages,
            'current_page': page,
            'per_page': per_page,
            'has_next': pagination.has_next,
            'has_prev': pagination.has_prev
        }

    @staticmethod
    def update_order_status(
        order_id: str,
        status: str
    ) -> Tuple[Optional[Order], Optional[Dict[str, Any]]]:
        """
        Update order status

        Args:
            order_id: Order ID
            status: New status

        Returns:
            Tuple of (order, error_dict)
        """
        order = Order.query.get(order_id)

        if not order:
            return None, {'error': 'Order not found'}

        # Validate status
        valid_statuses = ['pending', 'processing', 'shipped', 'delivered', 'cancelled']
        if status not in valid_statuses:
            return None, {'error': f'Invalid status. Must be one of: {", ".join(valid_statuses)}'}

        try:
            order.status = status

            # Update timestamps based on status
            if status == 'shipped':
                order.shipped_at = datetime.utcnow()
                order.fulfillment_status = 'fulfilled'
            elif status == 'delivered':
                order.delivered_at = datetime.utcnow()
                order.fulfillment_status = 'fulfilled'
            elif status == 'cancelled':
                order.fulfillment_status = 'cancelled'

            order.updated_at = datetime.utcnow()
            db.session.commit()

            return order, None

        except Exception as e:
            db.session.rollback()
            return None, {'error': f'Failed to update order status: {str(e)}'}

    @staticmethod
    def update_tracking_info(
        order_id: str,
        tracking_number: str = None,
        tracking_url: str = None
    ) -> Tuple[Optional[Order], Optional[Dict[str, Any]]]:
        """
        Update order tracking information

        Args:
            order_id: Order ID
            tracking_number: Tracking number
            tracking_url: Tracking URL

        Returns:
            Tuple of (order, error_dict)
        """
        order = Order.query.get(order_id)

        if not order:
            return None, {'error': 'Order not found'}

        try:
            if tracking_number:
                order.tracking_number = tracking_number
            if tracking_url:
                order.tracking_url = tracking_url

            order.updated_at = datetime.utcnow()
            db.session.commit()

            return order, None

        except Exception as e:
            db.session.rollback()
            return None, {'error': f'Failed to update tracking info: {str(e)}'}

    @staticmethod
    def get_all_users(
        page: int = 1,
        per_page: int = 20,
        role: str = None,
        is_active: bool = None
    ) -> Dict[str, Any]:
        """
        Get all users with filters and pagination

        Args:
            page: Page number
            per_page: Items per page
            role: Filter by role
            is_active: Filter by active status

        Returns:
            Dictionary with users and pagination info
        """
        query = User.query

        # Apply filters
        if role:
            query = query.filter(User.role == role)

        if is_active is not None:
            query = query.filter(User.is_active == is_active)

        # Order by created_at descending
        query = query.order_by(User.created_at.desc())

        # Paginate
        pagination = query.paginate(
            page=page,
            per_page=per_page,
            error_out=False
        )

        return {
            'users': [user.to_dict() for user in pagination.items],
            'total': pagination.total,
            'pages': pagination.pages,
            'current_page': page,
            'per_page': per_page,
            'has_next': pagination.has_next,
            'has_prev': pagination.has_prev
        }

    @staticmethod
    def update_user(
        user_id: str,
        **kwargs
    ) -> Tuple[Optional[User], Optional[Dict[str, Any]]]:
        """
        Update user details

        Args:
            user_id: User ID
            **kwargs: Fields to update (is_active, role)

        Returns:
            Tuple of (user, error_dict)
        """
        user = User.query.get(user_id)

        if not user:
            return None, {'error': 'User not found'}

        try:
            # Update allowed fields
            allowed_fields = ['is_active', 'role']

            for field, value in kwargs.items():
                if field in allowed_fields and value is not None:
                    # Validate role
                    if field == 'role' and value not in ['customer', 'admin']:
                        return None, {'error': 'Invalid role. Must be customer or admin'}

                    setattr(user, field, value)

            user.updated_at = datetime.utcnow()
            db.session.commit()

            return user, None

        except Exception as e:
            db.session.rollback()
            return None, {'error': f'Failed to update user: {str(e)}'}

File: frontend/src/services/adminService.ts
import api from './api'
import { Product, ProductFormData, ProductFilters, ProductListResponse } from '@/types/product.types'

export const adminService = {
  // ==================== Dashboard ====================

  async getDashboardStats() {
    const response = await api.get('/admin/dashboard')
    return response.data
  },

  // ==================== Product Management ====================

  async getProducts(filters: ProductFilters = {}): Promise<ProductListResponse> {
    const params = new URLSearchParams()

    if (filters.page) params.append('page', filters.page.toString())
    if (filters.per_page) params.append('per_page', filters.per_page.toString())
    if (filters.category_id) params.append('category_id', filters.category_id)
    if (filters.search) params.append('search', filters.search)
    if (filters.in_stock_only !== undefined) params.append('in_stock_only', filters.in_stock_only.toString())
    if (filters.is_active !== undefined) params.append('is_active', filters.is_active.toString())
    if (filters.is_featured !== undefined) params.append('is_featured', filters.is_featured.toString())
    if (filters.sort_by) params.append('sort_by', filters.sort_by)
    if (filters.sort_order) params.append('sort_order', filters.sort_order)

    const response = await api.get(`/admin/products?${params.toString()}`)
    return response.data
  },

  async getProduct(productId: string): Promise<Product> {
    const response = await api.get(`/admin/products/${productId}`)
    return response.data.product
  },

  async createProduct(data: ProductFormData): Promise<Product> {
    const formData = new FormData()

    formData.append('name', data.name)
    formData.append('price', data.price.toString())

    if (data.description) formData.append('description', data.description)
    if (data.category_id) formData.append('category_id', data.category_id)
    formData.append('stock_quantity', data.stock_quantity.toString())
    formData.append('low_stock_threshold', data.low_stock_threshold.toString())
    if (data.sku) formData.append('sku', data.sku)
    formData.append('is_featured', data.is_featured.toString())
    formData.append('is_active', data.is_active.toString())

    // Add main image
    if (data.image) {
      formData.append('image', data.image)
    }

    // Add additional images
    if (data.additional_images && data.additional_images.length > 0) {
      data.additional_images.forEach((file) => {
        formData.append('additional_images[]', file)
      })
    }

    // Add ingredients and allergens as JSON
    if (data.ingredients && data.ingredients.length > 0) {
      formData.append('ingredients', JSON.stringify(data.ingredients))
    }
    if (data.allergens && data.allergens.length > 0) {
      formData.append('allergens', JSON.stringify(data.allergens))
    }

    const response = await api.post('/admin/products', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })

    return response.data.product
  },

  async updateProduct(productId: string, data: Partial<ProductFormData>): Promise<Product> {
    const formData = new FormData()

    if (data.name) formData.append('name', data.name)
    if (data.price !== undefined) formData.append('price', data.price.toString())
    if (data.description) formData.append('description', data.description)
    if (data.category_id) formData.append('category_id', data.category_id)
    if (data.stock_quantity !== undefined) formData.append('stock_quantity', data.stock_quantity.toString())
    if (data.low_stock_threshold !== undefined) formData.append('low_stock_threshold', data.low_stock_threshold.toString())
    if (data.sku) formData.append('sku', data.sku)
    if (data.is_featured !== undefined) formData.append('is_featured', data.is_featured.toString())
    if (data.is_active !== undefined) formData.append('is_active', data.is_active.toString())

    // Add main image
    if (data.image) {
      formData.append('image', data.image)
    }

    // Add additional images
    if (data.additional_images && data.additional_images.length > 0) {
      data.additional_images.forEach((file) => {
        formData.append('additional_images[]', file)
      })
    }

    // Add ingredients and allergens as JSON
    if (data.ingredients) {
      formData.append('ingredients', JSON.stringify(data.ingredients))
    }
    if (data.allergens) {
      formData.append('allergens', JSON.stringify(data.allergens))
    }

    const response = await api.put(`/admin/products/${productId}`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })

    return response.data.product
  },

  async updateStock(productId: string, stockQuantity: number): Promise<Product> {
    const response = await api.put(`/admin/products/${productId}/stock`, {
      stock_quantity: stockQuantity,
    })
    return response.data.product
  },

  async deleteProduct(productId: string, hardDelete = false): Promise<void> {
    const params = hardDelete ? '?hard=true' : ''
    await api.delete(`/admin/products/${productId}${params}`)
  },

  async uploadProductImages(productId: string, images: File[]): Promise<Product> {
    const formData = new FormData()

    images.forEach((file) => {
      formData.append('images[]', file)
    })

    const response = await api.post(`/admin/products/${productId}/images`, formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    })

    return response.data.product
  },

  async getLowStockProducts(threshold?: number): Promise<Product[]> {
    const params = threshold ? `?threshold=${threshold}` : ''
    const response = await api.get(`/admin/products/low-stock${params}`)
    return response.data.products
  },

  // ==================== Order Management ====================

  async getOrders(filters: any = {}) {
    const params = new URLSearchParams()

    if (filters.page) params.append('page', filters.page.toString())
    if (filters.per_page) params.append('per_page', filters.per_page.toString())
    if (filters.status) params.append('status', filters.status)
    if (filters.payment_status) params.append('payment_status', filters.payment_status)
    if (filters.start_date) params.append('start_date', filters.start_date)
    if (filters.end_date) params.append('end_date', filters.end_date)

    const response = await api.get(`/admin/orders?${params.toString()}`)
    return response.data
  },

  async updateOrderStatus(orderId: string, status: string) {
    const response = await api.put(`/admin/orders/${orderId}/status`, { status })
    return response.data.order
  },

  async updateOrderTracking(orderId: string, trackingNumber: string, trackingUrl?: string) {
    const response = await api.put(`/admin/orders/${orderId}/tracking`, {
      tracking_number: trackingNumber,
      tracking_url: trackingUrl,
    })
    return response.data.order
  },

  // ==================== User Management ====================

  async getUsers(filters: any = {}) {
    const params = new URLSearchParams()

    if (filters.page) params.append('page', filters.page.toString())
    if (filters.per_page) params.append('per_page', filters.per_page.toString())
    if (filters.role) params.append('role', filters.role)
    if (filters.is_active !== undefined) params.append('is_active', filters.is_active.toString())

    const response = await api.get(`/admin/users?${params.toString()}`)
    return response.data
  },

  async updateUser(userId: string, data: { is_active?: boolean; role?: string }) {
    const response = await api.put(`/admin/users/${userId}`, data)
    return response.data.user
  },
}

export default adminService
